import_models_to_verta/mlflow_databricks_bulk_import_example.ipynb

```diff
@@ -0,0 +1,341 @@
+    "        if not rm_versions:\n",
+    "            failed_model_imports.append(rm)\n",
+    "            model_messages[rm].append(f\"unable to find any model versions for {rm.name}.  Skipping import.\")\n",
+    "            continue\n",
+    "    except Exception as err:\n",
+    "        model_messages[rm].append(f\"failed to fetch versions for registered model \\\"{rm.name}\\\". Skipping import. Error: {err}\")\n",
+    "        failed_model_imports.append(rm)\n",
+    "        continue\n",
+    "\n",
+    "    successful_versions = list()\n",
+    "    for version in rm_versions:\n",
+    "        try:\n",
+    "            verta_version = verta_rm.create_version(\n",
+    "                name=str(version.version),\n",
+    "                attrs={\n",
+    "                    \"er_id\": version.run_id,\n",
+    "                    \"mlflow_source\": version.source,\n",
+    "                    \"mlflow_user_id\": version.user_id,\n",
+    "                    \"mlflow_run_link\": version.run_link,\n",
+    "                    \"mlflow_creation_time\": version.creation_timestamp,\n",
+    "                    \"mlflow_last_updated_time\": version.last_updated_timestamp,\n",
+    "                    \"mlflow_status\": version.status,\n",
+    "                    \"mlflow_current_stage\": version.current_stage,\n",
+    "                    \"mlflow_tags\": version.tags,\n",
+    "                },\n",
+    "                labels=[\"mlflow_import\"],\n",
+    "            )\n",
+    "            model_messages[rm].append(f\"successfully created version {version.version}\")\n",
+    "        except Exception as err:\n",
+    "            model_messages[rm].append(f\"failed to create model version in Verta for {rm.name} - version: {version.version} due to {err}\")\n",
+    "            continue\n",
+    "\n",
+    "        # download artifacts\n",
+    "        try:\n",
+    "            outpath = mlflow.artifacts.download_artifacts(run_id=version.run_id)\n",
+    "        except Exception as err:\n",
+    "            model_messages[rm].append(f\"unable to download artifacts from {rm.name} - version run id; {version.run_id}\")\n",
+    "            outpath = None\n",
+    "\n",
+    "        if outpath:\n",
+    "            for file_name in os.listdir(outpath):\n",
+    "                if file_name not in protected_file_names:\n",
+    "                    verta_version.log_artifact(file_name, os.path.join(outpath, file_name))\n",
+    "                    model_messages[rm].append(f\"artifact logged in Verta: {file_name}\")\n",
+    "                else:\n",
+    "                    model_messages[rm].append(f\"cannot upload artifact {file_name} for {rm.name} as the file name conflicts with a protected file name in Verta\")\n",
+    "        else:\n",
+    "            model_messages[rm].append(f\"no artifacts found for {rm.name} - {version.version}\")\n",
```

Comment:
Isn't this redundant (and maybe incorrect)? This `else` looks like it's for `outpath == None`, and we already capture a message (`f\"unable to download artifacts from {rm.name} - version run id; {version.run_id}\"`) for this exact condition in the above `except` block.

Reply:
I think it's still inaccurate, because this `else` block _also_ gets executed if the exception above was thrown (because `outpath = None`. So if `download_artifacts()` encounters any exception—even if it's not due to no artifacts being found—we end up logging "no artifacts found" here.

> The MLflow API in use here could possibly return empty results without raising any exception

Also are you certain about this? I haven't run the MLflow client myself, but [the docs for `mlflow.artifacts.download_artifacts()`](https://mlflow.org/docs/latest/python_api/mlflow.artifacts.html#mlflow.artifacts.download_artifacts) say that it returns `str`, not `Optional[str]`.